File: GoogleNavigation/GoogleNavXCFrameworkDemos/Utils/NavDemoSwitch.m
/*
 * Copyright 2023 Google LLC. All rights reserved.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 * ANY KIND, either express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#import "GoogleNavXCFrameworkDemos/Utils/NavDemoSwitch.h"
#import <UIKit/UIKit.h>

@implementation NavDemoSwitch {
  UILabel *_label;
  UIColor *_textColor;
  UISwitch *_control;
}

+ (instancetype)switchWithLabel:(NSString *)label
                         target:(nullable id)target
                       selector:(nullable SEL)selector {
  return [self switchWithLabel:label initialState:NO target:target selector:selector];
}

+ (instancetype)switchWithLabel:(NSString *)label
                   initialState:(BOOL)initialState
                         target:(nullable id)target
                       selector:(nullable SEL)selector {
  return [[self alloc] initWithLabel:label
                           textColor:UIColor.labelColor
                        initialState:initialState
                              target:target
                            selector:selector];
}

- (instancetype)initWithLabel:(NSString *)label
                    textColor:(UIColor *)textColor
                 initialState:(BOOL)initialState
                       target:(nullable id)target
                     selector:(nullable SEL)selector {
  self = [super initWithFrame:CGRectZero];
  if (self) {
    _enabled = YES;
    _textColor = textColor;

    _label = [[UILabel alloc] initWithFrame:CGRectZero];
    _label.translatesAutoresizingMaskIntoConstraints = NO;
    _label.text = label;
    _label.textColor = textColor;
    _label.font = [UIFont systemFontOfSize:12];
    [self addArrangedSubview:_label];

    _control = [[UISwitch alloc] initWithFrame:CGRectZero];
    _control.tintColor = [UIColor blackColor];
    _control.on = initialState;
    _control.accessibilityIdentifier = [NSString stringWithFormat:@"switch - %@", label];
    if (target && selector) {
      SEL _Nonnull validSelector = (SEL _Nonnull)selector;
      [_control addTarget:target action:validSelector forControlEvents:UIControlEventValueChanged];
    }
    _control.translatesAutoresizingMaskIntoConstraints = NO;
    [self addArrangedSubview:_control];
    self.translatesAutoresizingMaskIntoConstraints = NO;
  }
  return self;
}

- (void)setEnabled:(BOOL)enabled {
  if (enabled == _enabled) {
    return;
  }
  _enabled = enabled;
  _control.enabled = enabled;
  _label.textColor = enabled ? _textColor : UIColor.lightGrayColor;
}

- (BOOL)isOn {
  return _control.on;
}

- (void)setOn:(BOOL)on {
  _control.on = on;
}

@end

File: GoogleNavigation-Swift/GoogleNavSwiftXCFrameworkDemos/Swift/Samples/Utils/NavDemoSwitch.swift
/// Copyright 2024 Google LLC. All rights reserved.
///
///
/// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
/// file except in compliance with the License. You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software distributed under
/// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
/// ANY KIND, either express or implied. See the License for the specific language governing
/// permissions and limitations under the License.

import UIKit

/// A switch with a title label on the left hand side and the switch on the right hand side.
class NavDemoSwitch: UIView {
  private let control = UISwitch()
  private var titleLabel: UILabel
  private var textColor: UIColor

  /// The state of the switch.
  var isOn: Bool {
    get { control.isOn }
    set { control.isOn = newValue }
  }

  /// Whether the switch is enabled.
  var isEnabled: Bool {
    didSet {
      guard oldValue != isEnabled else { return }
      control.isEnabled = isEnabled
      titleLabel.textColor = isEnabled ? textColor : .lightGray
    }
  }

  /// Creates a switch with the given options.
  ///
  /// - Parameters:
  ///   - title: The title which appears on the left hand side of the switch.
  ///   - textColor: The text color of the title. Defaults to `.black`.
  ///   - initialState: The initial state of the switch. Defaults to `false`.
  ///   - onValueChanged: The target/action pair for the `.valueChanged` event.
  init(
    title: String?, textColor: UIColor = .black, initialState: Bool = false,
    onValueChanged targetActionPair: TargetActionPair
  ) {
    self.titleLabel = MenuUIHelpers.makeLabel(text: title ?? "")
    self.textColor = textColor
    self.isEnabled = true

    super.init(frame: .zero)

    titleLabel.textColor = textColor

    let stackView = MenuUIHelpers.makeStackView(arrangedSubviews: [titleLabel, control])
    stackView.translatesAutoresizingMaskIntoConstraints = false
    stackView.distribution = .fill
    addSubview(stackView)

    control.isOn = initialState
    control.addTarget(
      targetActionPair.target, action: targetActionPair.action, for: .valueChanged)

    addConstraints([
      stackView.leadingAnchor.constraint(equalTo: leadingAnchor),
      stackView.trailingAnchor.constraint(equalTo: trailingAnchor),
      stackView.topAnchor.constraint(equalTo: topAnchor),
      stackView.bottomAnchor.constraint(equalTo: bottomAnchor),
    ])
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
}
